core/src/main/scala/org/allenai/pipeline/Pipeline.scala

```diff
@@ -0,0 +1,364 @@
+        new java.net.URI("http", s"${
+          url.getHost
+        }.s3.amazonaws.com", url.getPath, null)
+      case "file" =>
+        new java.net.URI(null, null, url.getPath, null)
+      case _ => url
+    }
+  }
+
+  def dryRun(outputDir: File, rawTitle: String, targets: Iterable[(String, Producer[_])] = persistedSteps.toList): Unit = {
+    val title = s"${
+      rawTitle.replaceAll("""\s+""", "-")
+    }-dryRun"
+    val workflowArtifact = new FileArtifact(new File(outputDir, s"$title.workflow.json"))
+    val workflow = Workflow.forPipeline(targets)
+    SingletonIo.json[Workflow].write(workflow, workflowArtifact)
+
+    val htmlArtifact = new FileArtifact(new File(outputDir, s"$title.html"))
+    SingletonIo.text[String].write(workflow.renderHtml, htmlArtifact)
+
+    val signatureArtifact = new FileArtifact(new File(outputDir, s"$title.signatures.json"))
+    val signatureFormat = Signature.jsonWriter
+    val signatures = targets.map { case (s, p) => signatureFormat.write(p.stepInfo.signature) }.toJson
+    signatureArtifact.write {
+      writer => writer.write(signatures.prettyPrint)
+    }
+
+    logger.info(s"Summary written to $outputDir")
+  }
+
+  protected[this] val steps =
+    scala.collection.mutable.Map.empty[String, PersistedProducer[_, _ <: Artifact]]
+}
+
+object Pipeline {
+  // Create a Pipeline that writes output to the given directory
+  def apply(rootDir: File) =
+    new Pipeline {
+      def rootOutputUrl = rootDir.toURI
+    }
+
+  def configured(cfg: Config) =
+    new ConfiguredPipeline {
+      val config = cfg
+    }
+}
+
+trait ConfiguredPipeline extends Pipeline {
+  val config: Config
+
+  override def rootOutputUrl =
+    config.get[String]("output.dir").map(s => new URI(s))
+      .getOrElse(new File(System.getProperty("user.dir")).toURI)
+
+  protected[this] def getStringList(key: String) =
+    config.get[String](key) match {
+      case Some(s) => List(s)
+      case None => config.get[Seq[String]](key) match {
+        case Some(sList) => sList
+        case None => List()
+      }
+    }
+
+  override def run(rawTitle: String) = {
+    val (targets, isRunOnly) =
+      getStringList("runOnly") match {
+        case seq if seq.nonEmpty =>
+          (getStepsByName(seq), true)
+        case _ =>
+          getStringList("runUntil") match {
+            case seq if seq.nonEmpty =>
+              (getStepsByName(seq), false)
+            case _ =>
+              (persistedSteps.toList, false)
+          }
+      }
+
+    config.get[Boolean]("dryRun") match {
+      case Some(true) =>
+        val outputDir = config.get[String]("dryRunOutput")
+          .getOrElse(System.getProperty("user.dir"))
+        dryRun(new File(outputDir), rawTitle, targets)
+        List()
+      case _ =>
+        if (isRunOnly) {
+          runOnly(rawTitle, targets)
+        } else {
+          runPipelineReturnResults(rawTitle, targets)
+        }
+    }
+  }
+
+  override def persist[T, A <: Artifact: ClassTag](
```

Comment:
As per email thread:

- Remove optionallyPersist() method, replace with persist()
- Behavior of persist() is to check config for an overridden output location
- If no config is found, behavior is to persist in the default location
- No longer necessary to set output.persist.XXX=true to enable persistence
- Can still disable persistence by setting output.persist.XXX=false in the config